src/Generics/SOP/Classes.hs

```diff
@@ -336,18 +376,90 @@ class HAp h => HSequence (h :: (k -> *) -> (l -> *)) where
   -- /Instances:/
   --
   -- @
-  -- 'hsequence'', 'Generics.SOP.NP.sequence'_NP'  :: ('SListI'  xs , 'Applicative' f) => 'Generics.SOP.NP.NP'  (f ':.:' g) xs  -> f ('Generics.SOP.NP.NP'  g xs )
-  -- 'hsequence'', 'Generics.SOP.NS.sequence'_NS'  :: ('SListI'  xs , 'Applicative' f) => 'Generics.SOP.NS.NS'  (f ':.:' g) xs  -> f ('Generics.SOP.NS.NS'  g xs )
+  -- 'hsequence'', 'Generics.SOP.NP.sequence'_NP'  :: ('Generics.SOP.Sing.SListI'  xs , 'Applicative' f) => 'Generics.SOP.NP.NP'  (f ':.:' g) xs  -> f ('Generics.SOP.NP.NP'  g xs )
+  -- 'hsequence'', 'Generics.SOP.NS.sequence'_NS'  :: ('Generics.SOP.Sing.SListI'  xs , 'Applicative' f) => 'Generics.SOP.NS.NS'  (f ':.:' g) xs  -> f ('Generics.SOP.NS.NS'  g xs )
   -- 'hsequence'', 'Generics.SOP.NP.sequence'_POP' :: ('SListI2' xss, 'Applicative' f) => 'Generics.SOP.NP.POP' (f ':.:' g) xss -> f ('Generics.SOP.NP.POP' g xss)
   -- 'hsequence'', 'Generics.SOP.NS.sequence'_SOP' :: ('SListI2' xss, 'Applicative' f) => 'Generics.SOP.NS.SOP' (f ':.:' g) xss -> f ('Generics.SOP.NS.SOP' g xss)
   -- @
   --
   hsequence' :: (SListIN h xs, Applicative f) => h (f :.: g) xs -> f (h g xs)
 
+-- ** Derived functions
+
 -- | Special case of 'hsequence'' where @g = 'I'@.
 hsequence :: (SListIN h xs, SListIN (Prod h) xs, HSequence h) => Applicative f => h f xs -> f (h I xs)
 hsequence = hsequence' . hliftA (Comp . fmap I)
 
 -- | Special case of 'hsequence'' where @g = 'K' a@.
 hsequenceK ::  (SListIN h xs, SListIN (Prod h) xs, Applicative f, HSequence h) => h (K (f a)) xs -> f (h (K a) xs)
 hsequenceK = hsequence' . hliftA (Comp . fmap K . unK)
+
+-- * Indexing into sums
+
+-- | A class for determining which choice in a sum-like structure
+-- a value represents.
+--
+class HIndex (h :: (k -> *) -> (l -> *)) where
+
+  -- | If 'h' is a sum-like structure representing a choice
+  -- between @n@ different options, and @x@ is a value of
+  -- type @h f xs@, then @'hindex' x@ returns a number between
+  -- @0@ and @n - 1@ representing the index of the choice
+  -- made by @x@.
+  --
+  -- /Instances:/
+  --
+  -- @
+  -- 'hindex', 'Generics.SOP.NS.index_NS'  :: 'Generics.SOP.NS.NS'  f xs -> Int
+  -- 'hindex', 'Generics.SOP.NS.index_SOP' :: 'Generics.SOP.NS.SOP' f xs -> Int
+  -- @
+  --
+  -- /Examples:/
+  --
+  -- >>> hindex (S (S (Z (I False))))
+  -- 2
+  -- >>> hindex (Z (K ()))
+  -- 0
+  -- >>> hindex (SOP (S (Z (I True :* I 'x' :* Nil))))
+  -- 1
+  --
+  -- @since 0.2.4.0
+  --
+  hindex :: h f xs -> Int
```

Comment:
Came to mind that `NS (K ()) xs` and `Fin (length xs) ≈ Int` are isomorphic. I'm not sure I needed (or whether it's useful) to have

```haskell
hunindex :: SList xs => Int -> (forall a. f a) -> h f xs
hunindex n x = hapInjs (hpure x) !! n

hunindexc :: (SList xs, All c xs) => Proxy c => Int -> (forall a. c a => f a) -> h f xs
hunindexc n c x = hapInjs (hcpure c x) !! n
```

directly available?